File: LeetCode/remote/surrounded_regions.cs
//  https://leetcode.com/problems/surrounded-regions/
//
//   Given a 2D board containing 'X' and 'O', capture all regions surrounded by 'X'.
//
//   A region is captured by flipping all 'O's into 'X's in that surrounded region.
//
//   For example,
//
//   X X X X
//   X O O X
//   X X O X
//   X O X X
//
//   After running your function, the board should be:
//
//   X X X X
//   X X X X
//   X X X X
//   X O X X
//

using System;
using System.Collections.Generic;

public class Solution {
    public void Solve(char[,] board) {
        var n = board.GetLength(0);
        var m = board.GetLength(1);
        var borderReplacements = new Dictionary<char, char> { {'O', 'Y'} };
        for (var i = 0; i < n; i++)
        {
            Fill(board, i, m - 1, borderReplacements);
            Fill(board, i, 0, borderReplacements);
        }

        for (var i = 0; i < m; i++)
        {
            Fill(board, n - 1, i, borderReplacements);
            Fill(board, 0, i, borderReplacements);
        }

        var finalReplacements = new Dictionary<char, char> { {'O', 'X'} };
        for (var i = 0; i < n; i++)
        {
            for (var j = 0; j < m; j++)
            {
                Fill(board, i, j, finalReplacements);
            }
        }

        borderReplacements = new Dictionary<char, char> { {'Y', 'O'} };
        for (var i = 0; i < n; i++)
        {
            Fill(board, i, m - 1, borderReplacements);
            Fill(board, i, 0, borderReplacements);
        }

        for (var i = 0; i < m; i++)
        {
            Fill(board, n - 1, i, borderReplacements);
            Fill(board, 0, i, borderReplacements);
        }
    }

    public void Fill(char[,] board, int line, int col, Dictionary<char, char> replacements)
    {
        var stack = new Stack<Tuple<int, int>>();
        stack.Push(Tuple.Create(line, col));

        while (stack.Count > 0)
        {
            var current = stack.Pop();
            line = current.Item1;
            col = current.Item2; 
            if (!replacements.ContainsKey(board[line, col]))
            {
                continue;
            }

            board[line, col] = replacements[board[line, col]];
            for (var a = Math.Max(0, line - 1); a <= Math.Min(board.GetLength(0) - 1, line + 1); a++)
            {
                for (var b = Math.Max(0, col - 1); b <= Math.Min(board.GetLength(1) - 1, col + 1); b++)
                {
                    if (a != b)
                    {
                        stack.Push(Tuple.Create(a, b));
                    }
                }
            }
        }
    }

    public void Print(char[,] board)
    {
        for (var i = 0; i < board.GetLength(0); i++)
        {
            for (var j = 0; j < board.GetLength(1); j++)
            {
                Console.Write("{0} ", board[i, j]);
            }

            Console.WriteLine();
        }

        Console.WriteLine();
    }

    static void Main()
    {
        var s = new Solution();
        var a = new char[4, 4] { 
            { 'X', 'X', 'X', 'X' },
            { 'X', 'O', 'O', 'X' },
            { 'X', 'X', 'O', 'X' },
            { 'X', 'O', 'X', 'X' },
        };

        s.Print(a);
        s.Solve(a);
        s.Print(a);
        
        a = new char[2, 2] { 
            { 'O', 'O' },
            { 'O', 'O' }
        };

        s.Print(a);
        s.Solve(a);
        s.Print(a);
    }
}

File: LeetCode/remote/surrounded_regions.rb
#   https://leetcode.com/problems/surrounded-regions/
#    Given a 2D board containing 'X' and 'O', capture all regions surrounded by 'X'.
#
#    A region is captured by flipping all 'O's into 'X's in that surrounded region.
#
#    Pretty cute solution but it blows up the stack.
#    Line 15: in `between?': stack level too deep (SystemStackError)
# @param {Character[][]} board
# @return {Void} Do not return anything, modify board in-place instead.
def solve(board)
    return if board.empty?
    (0...board.size)
    .to_a
    .product((0...board.first.size).to_a)
    .each { |l, c|
        next if board[l][c] == 'X'
        dfs(board, l, c)
    }
end

def dfs(board, line, column)
    return false if !line.between?(0, board.size - 1) || !column.between?(0, board.first.size - 1)
    return true if board[line][column] == 'X'
    board[line][column] = 'X'
    good = dfs(board, line - 1, column) &&
           dfs(board, line + 1, column) &&
           dfs(board, line, column - 1) &&
           dfs(board, line, column + 1)
    board[line][column] = 'O' unless good
    return good
end

# It's evolution baby
def solve(board)
    (0...board.size).each { |i|
        (0...board[0].size).each { |j|
            fill(board, i, j)
        }
    }
end

def fill(board, i, j)
    return false if !i.between?(0, board.size - 1) || !j.between?(0, board[0].size - 1)
    return true if board[i][j] == 'X'
    board[i][j] = 'X'
    board[i][j] = '0' if !fill(board, i - 1, j) ||
                         !fill(board, i + 1, j) ||
                         !fill(board, i, j - 1) ||
                         !fill(board, i, j + 1)
end

def solve(board)
    (0...board.size).each { |i|
        (0...board[0].size).each { |j|
            fill(board, i, j) unless board[i][j] == 'X'
        }
    }
end

def fill(board, i, j)
    stack = [[[i, j], [i, j]]]
    should_fill = true
    while stack.any?
        frontier = stack.pop
        current = c(frontier)
        parent = p(frontier)
        if !should_fill
            board[parent[0]][parent[1]] = 'O'
        else
            board[current[0]][current[1]] = 'X'
            (current[0] - 1).upto(current[0] + 1).each { |i|
                (current[1] - 1).upto(current[1] + 1).each { |j|
                    next if i == j
                    should_fill &&= i.between?(0, board.size - 1) && j.between?(0, board[0].size - 1) 
                    next if !should_fill
                    stack.push([[i, j], current])
                }
            }      
        end
        
        board[current[0]][current[1]] = 'O' if !should_fill
    end
end

def p(x)
    return x[1]
end

def c(x)
    return x[0]
end
